src/Domain/DoctrineDonationRepository.php

```diff
@@ -90,10 +87,28 @@ public function allocateMatchFunds(Donation $donation): string
             $amountNewlyMatched = bcadd($amountNewlyMatched, $newWithdrawalAmount, 2);
         }
 
-        $this->getEntityManager()->flush(); // Flush `$newWithdrawals` if any.
+        try {
+            // Flush `$newWithdrawals` if any and updates to DB copies of CampaignFundings.
+            $this->getEntityManager()->flush();
+        } catch (\Throwable $exception) {
+            $this->getLogger()->error(sprintf(
+                'Failed to flush DB for new withdrawals so donation UUID %s did not get matching: %s',
+                $donation->getUuid()->toString(),
+                $exception->getMessage(),
+            ));
+
+            // todo pick one
+            $this->matchingAdapter->releaseNewlyAllocatedFunds();
+            //$this->reverseMatchingAllocations($donation, $newWithdrawals, $likelyAvailableFunds);
+
+            // Ensure nothing later tries to persist the pending Donation or CampaignFunding changes.
+            $this->getEntityManager()->clear();
+
+            return '0.0';
```

Comment:
I think if we're going to clear the EM then we should throw effectively straight back to the UI. There's no reason for a caller to assume that this might clear the EM so I think it might be hard to discover that or understand why changes aren't getting persisted if we don't remember that this clear line is here.

Also probably better to close the em instead of just clearing it, so any attempt to flush after this would throw instead of letting us think we'd successfully persisted changes.

Reply:
Yep, agreed – this should be done in e9d51a4b734c01357d31a7fa537cda242124f5dc